### DISCUSSION TITLE
Potential resource leak in SupportBean and LogBean

### DISCUSSION THREAD

**Discussion by alexsuter** [axonivy/engine-cockpit#361] [Q&A]
Falls ich den Code richtig verstehe. Dies könnte die ganze Engine verblasen? Hier wird der ganze Stream bereits konsumiert?!
https://github.com/axonivy/engine-cockpit/blob/56d3d2b6cc4bd750727ce79d2969519db250d953/engine-cockpit/src/ch/ivyteam/enginecockpit/system/SupportBean.java#L42

Zudem ist nicht sichergestellt, dass "out" geschlossen wird (try-with-resources).

Weiter unten im Code, werden auch alle Log files von einem Ort in ein Temp-Directory kopiert. Sprich, wenn man schon recht am Limit ist mit der Disk Space, dann könnte das zusätzlich zu einem Out-Of-Diskspace führen.
https://github.com/axonivy/engine-cockpit/blob/56d3d2b6cc4bd750727ce79d2969519db250d953/engine-cockpit/src/ch/ivyteam/enginecockpit/system/SupportBean.java#L61

Dieselbe Problematik besteht auch im LogBean:
https://github.com/axonivy/engine-cockpit/blob/56d3d2b6cc4bd750727ce79d2969519db250d953/engine-cockpit/src/ch/ivyteam/enginecockpit/monitor/LogBean.java#L82

@ivy-lli @ivy-cst @jackra1n

**Comment by ivy-lli** [comment]
At least the OutputStream is closed in the LogBean 🙈 
The PrimeFaces StreamContent needs an InputStream, so it needs to be converted somehow. Except directly an InputStream is generated...

Maybe also a problem on the BrandingBean?

**Reply by alexsuter** [reply]
As far as I know, this is what piped stream are made for:

```java
var in = new PipedInputStream();
var out = new PipedOutputStream(in);
```

Another assumption:
But even if we would change this to piped streams, we still will have all data in memory. Because all data will be written to output stream first, after this has been finished, the stream will be given to primeface.

Writing to the stream must be made async or on demand. So if someone is STARTING to consume the stream, we need to start to read the files and write them to the stream.

Yes same problem here:
https://github.com/axonivy/engine-cockpit/blob/56d3d2b6cc4bd750727ce79d2969519db250d953/engine-cockpit/src/ch/ivyteam/enginecockpit/configuration/BrandingBean.java#L203

But maybe not that crucial, because log files can get huge.

**Reply by ivy-lli** [reply]
Never used a piped stream, but yes looks good.

And yes I think you're right
